File: backend/src/scripts/audit_git_course.ts
import { db } from '../firebase';

const EXPECTED_TITLES = [
  'Module 1: Introduction to Version Control, Git & GitHub',
  'Module 2: Installing Git and Initial Configuration',
  'Module 3: Git Repository Fundamentals',
  'Module 4: Basic Git Commands',
  'Module 5: Branching and Merging',
  'Module 6: GitHub Basics',
  'Module 7: Remote Repository Management',
  'Module 8: Git Collaboration',
  'Module 9: Advanced Git Commands',
  'Module 10: Git Internals',
  'Module 11: GitHub Features',
  'Module 12: Git Best Practices',
  'Module 13: Real-World Git Workflow',
  'Module 14: Git & GitHub Projects',
  'Module 15: Git & GitHub Interview Preparation'
];

const FOREIGN_KEYWORDS = {
  React: ['jsx', 'virtual dom', 'usestate', 'useeffect', 'redux toolkit'],
  Kubernetes: ['kubelet', 'minikube', 'kubernetes architecture', 'pvc', 'pv claim', 'persistentvolume'],
  Linux: ['systemd background', 'setfacl', 'sudoers visudo', 'chage password', 'monolithic kernel']
};

async function runAudit() {
  console.log('--- STARTING LIVE DATABASE AUDIT (audit_git_course.ts) ---');
  if (!db) {
    console.error('Firebase DB not initialized');
    process.exit(1);
  }

  const courseId = 'git-github-mastery-course-id';
  let passedCount = 0;

  try {
    const courseDoc = await db.collection('courses').doc(courseId).get();
    if (!courseDoc.exists) {
      console.error(`Course "${courseId}" not found in Firestore.`);
      process.exit(1);
    }
    const courseData = courseDoc.data() || {};
    console.log(`Course Found: "${courseData.title}" | ID: ${courseDoc.id}`);
    console.log(`Nested Modules: ${courseData.modules?.length || 0} modules`);

    for (let idx = 1; idx <= 15; idx++) {
      const expectedTitle = EXPECTED_TITLES[idx - 1];
      const modId = `git-mod-${idx}`;
      const topicId = `git-topic-${idx}`;
      const unitId = `git-unit-${idx}-notes`;

      const issues: string[] = [];
      const foreignContent: string[] = [];

      // 1. Verify modules collection document
      const modDoc = await db.collection('modules').doc(modId).get();
      if (!modDoc.exists) {
        issues.push(`Module document "${modId}" is missing in Firestore`);
      } else {
        const modData = modDoc.data() || {};
        if (modData.title !== expectedTitle) {
          issues.push(`Title mismatch: "${modData.title}" vs "${expectedTitle}"`);
        }
        if (modData.courseId !== courseId) {
          issues.push(`Course ID mapping incorrect: expected "${courseId}", got "${modData.courseId}"`);
        }
      }

      // 2. Verify lessons collection document
      const lessonDoc = await db.collection('lessons').doc(unitId).get();
      if (!lessonDoc.exists) {
        issues.push(`Lesson document "${unitId}" is missing in Firestore`);
      } else {
        const lessonData = lessonDoc.data() || {};
        if (lessonData.moduleId !== modId) {
          issues.push(`Lesson moduleId mapping incorrect: expected "${modId}", got "${lessonData.moduleId}"`);
        }
        if (lessonData.courseId !== courseId) {
          issues.push(`Lesson courseId mapping incorrect: expected "${courseId}", got "${lessonData.courseId}"`);
        }
        if (lessonData.type !== 'reading') {
          issues.push(`Lesson type should be "reading", got "${lessonData.type}"`);
        }

        // Check content keywords
        const content = (lessonData.content || '').toLowerCase();
        for (const [category, keywords] of Object.entries(FOREIGN_KEYWORDS)) {
          for (const kw of keywords) {
            if (content.includes(kw)) {
              foreignContent.push(`${category} (${kw})`);
            }
          }
        }
      }

      // 3. Verify nested course.modules array
      const nestedMod = courseData.modules?.find((m: any) => m.id === modId);
      if (!nestedMod) {
        issues.push(`Nested module "${modId}" not found in course document`);
      } else {
        if (nestedMod.title !== expectedTitle) {
          issues.push(`Nested title mismatch: "${nestedMod.title}" vs "${expectedTitle}"`);
        }
        const topic = nestedMod.topics?.[0];
        if (!topic) {
          issues.push('Nested topic is missing');
        } else {
          if (topic.id !== topicId) {
            issues.push(`Nested topic ID mismatch: expected "${topicId}", got "${topic.id}"`);
          }
          if (nestedMod.topics.length > 1) {
            issues.push(`Multiple topics found: expected 1, got ${nestedMod.topics.length}`);
          }
          
          const unit = topic.learningUnits?.[0];
          if (!unit) {
            issues.push('Nested learning unit is missing');
          } else {
            if (unit.id !== unitId) {
              issues.push(`Nested unit ID mismatch: expected "${unitId}", got "${unit.id}"`);
            }
            if (unit.type !== 'Reading') {
              issues.push(`Nested unit type should be "Reading", got "${unit.type}"`);
            }
            if (topic.learningUnits.length > 1) {
              issues.push(`Multiple learning units found: expected 1, got ${topic.learningUnits.length}`);
            }
          }
        }
      }

      const passed = issues.length === 0 && foreignContent.length === 0;
      if (passed) passedCount++;

      console.log(`\nModule ${idx}: "${expectedTitle}"`);
      console.log(`  Live DB Status:   ${modDoc.exists && lessonDoc.exists ? 'PASSED' : 'FAILED'}`);
      console.log(`  Relationships:    ${issues.some(i => i.includes('ID') || i.includes('mapping')) ? 'INCORRECT' : 'CORRECT'}`);
      console.log(`  Foreign Content:  ${foreignContent.length > 0 ? foreignContent.join(', ') : 'None'}`);
      if (issues.length > 0) {
        console.log(`  Issues Detected:`);
        issues.forEach(iss => console.log(`    - ${iss}`));
      }
    }

    console.log(`\n--- SUMMARY (audit_git_course.ts) ---`);
    console.log(`Total Modules Audited: 15`);
    console.log(`Passed Modules:  ${passedCount}/15`);
    console.log(`Overall Result:  ${passedCount === 15 ? 'PASSED' : 'FAILED'}`);

  } catch (error) {
    console.error('Audit failed:', error);
    process.exit(1);
  }
  process.exit(0);
}

runAudit();

File: backend/src/scripts/audit_git_course.py
import re
import sys

# Define expected module titles
EXPECTED_TITLES = [
    'Module 1: Introduction to Version Control, Git & GitHub',
    'Module 2: Installing Git and Initial Configuration',
    'Module 3: Git Repository Fundamentals',
    'Module 4: Basic Git Commands',
    'Module 5: Branching and Merging',
    'Module 6: GitHub Basics',
    'Module 7: Remote Repository Management',
    'Module 8: Git Collaboration',
    'Module 9: Advanced Git Commands',
    'Module 10: Git Internals',
    'Module 11: GitHub Features',
    'Module 12: Git Best Practices',
    'Module 13: Real-World Git Workflow',
    'Module 14: Git & GitHub Projects',
    'Module 15: Git & GitHub Interview Preparation'
]

# Foreign track keywords
FOREIGN_KEYWORDS = {
    'React': ['jsx', 'virtual dom', 'usestate', 'useeffect', 'redux toolkit'],
    'Kubernetes': ['kubelet', 'minikube', 'kubernetes architecture', 'pvc', 'pv claim', 'persistentvolume'],
    'Linux': ['systemd background', 'setfacl', 'sudoers visudo', 'chage password', 'monolithic kernel']
}

def main():
    sys.stdout.reconfigure(encoding='utf-8')
    
    frontend_path = r"C:\Users\devis\LMS-Platform\frontend\src\data\gitCourseFullData.ts"
    backend_path = r"C:\Users\devis\LMS-Platform\backend\src\services\course\gitSyllabusData.ts"
    
    print("--- STARTING STATIC AUDIT (audit_git_course.py) ---")
    
    with open(frontend_path, "r", encoding="utf-8") as f:
        fe_content = f.read()
        
    with open(backend_path, "r", encoding="utf-8") as f:
        be_content = f.read()
        
    results = []
    clean_count = 0
    
    for idx, expected_title in enumerate(EXPECTED_TITLES, 1):
        mod_id = f"git-mod-{idx}"
        topic_id = f"git-topic-{idx}"
        unit_id = f"git-unit-{idx}-notes"
        
        fe_status = "Missing"
        be_status = "Missing"
        relationship_status = "Correct"
        foreign_content = []
        issues = []
        
        # 1. Inspect frontend file
        # Search for module block
        fe_mod_match = re.search(rf"id:\s*'{mod_id}'(.*?)(?=\s+id:\s*'git-mod-|\Z)", fe_content, re.DOTALL)
        fe_module_text = ""
        if fe_mod_match:
            fe_module_text = fe_mod_match.group(1)
            
            # Check title
            if expected_title in fe_module_text:
                fe_status = "Correct"
            else:
                fe_status = "Mixed"
                issues.append("Title mismatch in frontend")
                
            # Check ID relationships
            if f"id: 'git-topic-{idx}'" not in fe_module_text:
                relationship_status = "Incorrect"
                issues.append(f"Topic ID mismatch (expected 'git-topic-{idx}')")
            if f"id: 'git-unit-{idx}-notes'" not in fe_module_text:
                relationship_status = "Incorrect"
                issues.append(f"Unit ID mismatch (expected 'git-unit-{idx}-notes')")
            if "type: 'Reading'" not in fe_module_text:
                issues.append("Unit type is not 'Reading'")
        else:
            issues.append(f"Module {mod_id} not found in frontend file")
            
        # 2. Inspect backend file
        be_mod_match = re.search(rf"{idx}:\s*`(.*?)(?=`\s*,\s*\d+:\s*`|\Z)", be_content, re.DOTALL)
        be_module_text = ""
        if be_mod_match:
            be_module_text = be_mod_match.group(1)
            be_status = "Correct"
        else:
            issues.append(f"Module {idx} content not found in backend file")
            be_status = "Missing"
            
        # 3. Check for foreign content keywords
        combined_text = (fe_module_text + " " + be_module_text).lower()
        for category, keywords in FOREIGN_KEYWORDS.items():
            for kw in keywords:
                if kw in combined_text:
                    foreign_content.append(f"{category} ({kw})")
                    
        is_clean = (fe_status == "Correct" and 
                    be_status == "Correct" and 
                    relationship_status == "Correct" and 
                    len(foreign_content) == 0 and 
                    len(issues) == 0)
        
        if is_clean:
            clean_count += 1
            
        results.append({
            "num": idx,
            "title": expected_title,
            "fe_status": fe_status,
            "be_status": be_status,
            "relationship": relationship_status,
            "foreign": foreign_content,
            "issues": issues
        })
        
    print("\n--- AUDIT VERIFICATION REPORT (audit_git_course.py) ---")
    for r in results:
        print(f"\nModule {r['num']}: \"{r['title']}\"")
        print(f"  Frontend Status:     {r['fe_status']}")
        print(f"  Backend Status:      {r['be_status']}")
        print(f"  ID Relationships:    {r['relationship']}")
        print(f"  Foreign Content:     {r['foreign'] if r['foreign'] else 'None'}")
        if r['issues']:
            print("  Issues:")
            for iss in r['issues']:
                print(f"    - {iss}")
                
    print("\n--- SUMMARY (audit_git_course.py) ---")
    print(f"Audited Modules: 15")
    print(f"Passed Modules:  {clean_count}/15")
    print(f"Overall Result:  {'PASSED' if clean_count == 15 else 'FAILED'}")

if __name__ == "__main__":
    main()
